File: 477_TotalHammingDistance.js
/**
 * @param {number[]} nums
 * @return {number}
 */
//Runtime: 64 ms, faster than 86.08% of JavaScript online submissions for Total Hamming Distance.
//Memory Usage: 37.4 MB, less than 100.00% of JavaScript online submissions for Total Hamming Distance.
var totalHammingDistance = function(nums) {
    var res=0;
    for(var i=0;i<32;i++){
        var count=0;
        for(j=0;j<nums.length;j++){
            count+=(nums[j]>>i)&1
        };
        res+=count*(nums.length-count);
    }
    return res;
    
};
var inlist=[4,14,2];
//console.log(inlist);
console.log(totalHammingDistance(inlist));
console.log('-----------the end!----------------');

File: 477_TotalHammingDistance.py
class Solution(object):
    def totalHammingDistance(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        #Runtime: 372 ms, faster than 79.60% of Python online submissions for Total Hamming Distance.
        #Memory Usage: 12.8 MB, less than 50.00% of Python online submissions for Total Hamming Distance.
        res=0
        for i in range(32):
            # for each bit
            count=0
            for v in nums:
                count+= (v>>i)&1
            # the count in this bit: multiplication of  the 1 nubmers and  0 numbers
            res+= count*(len(nums)-count)
        return res
    def totalHammingDistance1(self, nums: list[int]) -> int:
        # Time Limit Exceeded: 35 / 46 test cases passed.
        # force solution: time: o(n**2) space: o(1)
        res = 0
        for i in range(len(nums)):
            for j in range(i+1, len(nums)):
                if nums[i] != nums[j]:
                    res += self.hammingDistance(nums[i], nums[j])
        return res

    def hammingDistance(self, n1, n2):
        #print('n1: {} n2: {} dis: {}'.format(n1, n2, str(bin(n1 ^ n2))))
        return (str(bin(n1 ^ n2))).count('1')

    def totalHammingDistance(self, nums: list[int]) -> int:
        # Runtime: 695 ms, faster than 46.78% of Python3 online submissions for Total Hamming Distance.
        # Memory Usage: 15.6 MB, less than 28.07% of Python3 online submissions for Total Hamming Distance.
        # frequency logical solution: time: o(32* n) = o(n) space: o(1)
        res = 0
        # find 1  number and 0 number in i bit
        # the differece sum in i bit:  the combination of (1, 0) = ones * zeros
        for i in range(32):
            ones = 0
            for v in nums:
                # &1 : just get the most left bit
                ones += (v >> i) & 1
            res += ones * (len(nums)-ones)
        return res

if __name__ == '__main__':
    object = Solution()
    A= [4,14,2]

    print('---Start---')
    r = object.totalHammingDistance(A)
    print(r)
    print('---End---')
